### PR TITLE
gitlab CI support

### PR DESCRIPTION
**Pull request by swapnilmishra**
This PR is complementing the changes required for https://github.com/siddharthkp/bundlesize/issues/19 

1. Add `gitlab` as one of the `ci` variable value.
2. Tests for gitlab ci variable.
3. Adds .gitlab-ci.yml to run the test in gitlab CI.

gitlab CI output is attached below. For the CI test, I had to mirror the forked repo from github => gitlab and vice versa.

image

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -1,5 +1,9 @@
 let drone = require('./utils/drone')
 let repo, sha, event, commit_message, branch, ci
+// platform denotes code hosting provider i.e github, gitlab, bitbucket etc.
+// Had to introduce this variable as there are cases when CI is run on the same platform where code is hosted as those cases need to be handled differently.
+// Default value is github
+let platform = 'github';
```
Comment (siddharthkp) [inline index.js]: that's a risky default

File: test.js [hunk 2]
```diff
@@ -50,6 +50,10 @@ if (ci) {
       t.is(branch, real_branch)
     }
   })
+} else if (ci && platform === 'gitlab') {
```
Comment (siddharthkp) [inline test.js]: would you not want to test any of the other things 🤔

**Comment by joshghent** [comment]
Hey @siddharthkp, if I get this branch fixed up, could we get it merged so we can also merge https://github.com/siddharthkp/bundlesize/pull/180? 

Very keen to get bundlesize up and running at my company 👍

**Comment by swapnilmishra** [comment]
Hey @joshghent @siddharthkp I might get some-time this week to work or provide any support on it.

**Comment by HiddeRpl** [comment]
Year after and still no merge?

**Comment by siddharthkp** [comment]
@hidderpl would you like to contribute your time to take this pull request through?

**Comment by swapnilmishra** [comment]
@siddharthkp As I mentioned before I can work on this to take it to completion.
Would you be able to review the final set of changes then?
I will wait for your response before starting the work.

**Comment by siddharthkp** [comment]
I might not, cc @kuldeepkeshwar who takes care of this repo

**Comment by kuldeepkeshwar** [comment]
@swapnilmishra [predefined_variables](https://docs.gitlab.com/ee/ci/variables/predefined_variables.html) 👈 might be helpful

**Comment by swapnilmishra** [comment]
@kuldeepkeshwar Thanks for the link. The `GITLAB_CI` variable that I am using to identify if the CI is from gitlab is from that only. Would you also like to use any other variable here?

What I feel is we don't need to use anything else for gitlab CI because the reporting happens within the platform itself based on the exit code. This is because in the case of gitlab, the CI is integrated into the platform itself. This means we don't need detail like `repo`, `commit_message`, `sha` etc. The status badge is also provided by gitlab based on the CI status.

But please let me know if you would like me to make more changes. I would be happy to do them in order to close the PR. 👍

**Comment by swapnilmishra** [comment]
Just explaining the changes a bit more to give more context as the PR is quite old. 

So by default `bundlesize` works fine on gitlab CI but it just throws this warning message which is confusing. The message which is shown is from this line from `bundlesize` codebase https://github.com/siddharthkp/bundlesize/blob/master/src/api.js#L14. See [this image](https://user-images.githubusercontent.com/875450/32401250-9b5d3ad0-c10b-11e7-81e5-8c6b616f5d05.png) for the exact CI output.

At the moment the line which use to show warning(at the time of creating this PR) is commented out but whenever this change is reverted(mentioned in TODO) comment it is going to show the message again. So all we needed was a way to figure out that the `bundlesize` is running in gitlab CI and not show the warning.

**Comment by siddharthkp** [comment]
just curious, which version of bundlesize are you running?

**Comment by swapnilmishra** [comment]
I am using version `0.18.0` which still prints the warning message. And I guess that is because the changes related to commenting the warning are not yet released but are there in master.

**Comment by siddharthkp** [comment]
If you know the file, you could check it on unpkg: https://unpkg.com/browse/bundlesize@0.18.0/

**Comment by swapnilmishra** [comment]
Yes I can see the warn message in https://unpkg.com/browse/bundlesize@0.18.0/src/api.js at line no. 15

**Comment by swapnilmishra** [comment]
Are you planning to release the current changes(of master) in which the warn message is commented out([given here](https://github.com/siddharthkp/bundlesize/blob/master/src/api.js#L17)) in the next release?

I am asking because if the warn message is permanently going to be removed then we don't even need this change.

**Comment by swapnilmishra** [comment]
The output of gitlab CI with the current version can be seen here https://gitlab.com/swapnilmishra/test-ci/-/jobs/348930852

**Comment by kuldeepkeshwar** [comment]
@swapnilmishra IMO we should also capture other variables too.

All the information provided by `ci-env` is always present in the environment. `ci-env` is provide a uniformed way of accessing that information.

**Comment by swapnilmishra** [comment]
@kuldeepkeshwar Sure, I will add them. 👍

**Comment by swapnilmishra** [comment]
@kuldeepkeshwar 
Added all the CI variables for `gitlab` along with tests for each of them.
CI output can be viewed: https://gitlab.com/swapnilmishra/ci-env/-/jobs/354041324 and in the screenshot below.
Screenshot 2019-11-18 at 17 38 25

File: test.js [hunk 3]
```diff
@@ -118,6 +119,38 @@ if (ci) {
       t.is(branch, real_branch);
     }
   });
+} else if (ci && platform === "gitlab") {
+  test("ci is correctly set", t => {
```
Comment (kuldeepkeshwar) [inline test.js]: shouldn't we check for `ci`?
It seems it is obvious that if `ci` is `gitlab` then the `platform` is also `gitlab`.  @swapnilmishra can you confirm? if so then IMO we should put the check for `ci`

Comment (swapnilmishra) [reply test.js]: @kuldeepkeshwar 
Yes, that is true.
In that case, the condition will become `else if (ci === "gitlab") {` and then `t.is(ci, "gitlab");` test can be removed.

The reason I kept it like that was, to be in line with the other tests but can make this change.

Comment (kuldeepkeshwar) [reply test.js]: @swapnilmishra can you remove the `platform` check?
By removing the `platform` check test can fit the existing structure(get rid of if/else block).

Comment (swapnilmishra) [reply test.js]: @kuldeepkeshwar 
Maybe I understood it wrong but if we remove `platform` check and also remove `if else` condition how will we know when to run `gitlab` tests(which can only be run on gitlab) and when to just run `github` tests(which can only be run on github).

I can quickly make the changes if you can help me figuring out how do you think the conditions would look like. Right now it is:

```js
if (ci && platform === "github") {
 // github tests
} else if (ci && platform === "gitlab") {
 // gitlab tests
}
```

Now If we remove the platform and remove any if-else conditions both `gitlab` and `github` tests would be running.

Comment (kuldeepkeshwar) [reply test.js]: I was thinking something like 👇

```js
const test = require("ava");

const {
  buildUrl,
  jobUrl,
  repo,
  sha,
  event,
  commit_message,
  pull_request_number,
  branch,
  ci
} = require("./index");

if (ci) {
  console.log("values: ", {
    repo,
    sha,
    event,
    commit_message,
    pull_request_number,
    branch,
    ci
  });
 // ... e.g
  test("sha is set", t => {
    const real_sha =
      process.env.TRAVIS_PULL_REQUEST_SHA ||
      process.env.TRAVIS_COMMIT ||
      process.env.CIRCLE_SHA1 ||
      process.env.WERCKER_GIT_COMMIT ||
      process.env.DRONE_COMMIT ||
      process.env.GITHUB_SHA|| 
      process.env.CI_COMMIT_SHA; // add env variable here

    t.is(sha, real_sha);
  });

} else if (ci && platform === "gitlab") { // ... remove this block
  
} else {
  test.skip("These tests can only run in CI environments", t => t.pass());
}
```
this is how existing test are structured currently.

Comment (swapnilmishra) [reply test.js]: @kuldeepkeshwar 
Thanks for the explanation. I have made those changes now.

**Comment by kuldeepkeshwar** [comment]
@swapnilmishra thanks for pulling this off

**Pull request merged**